org.jacoco.core/src/org/jacoco/core/internal/analysis/MethodCoverageCalculator.java

```diff
@@ -105,18 +105,36 @@ private void applyMerges() {
 	}
 
 	private void applyReplacements() {
-		for (final Entry<AbstractInsnNode, Set<AbstractInsnNode>> entry : replacements
+		final HashMap<AbstractInsnNode, Instruction> replacedInstructions = new HashMap<AbstractInsnNode, Instruction>();
+		for (final Entry<AbstractInsnNode, Map<AbstractInsnNode, Collection<InstructionBranch>>> entry : replacements
 				.entrySet()) {
-			final Set<AbstractInsnNode> replacements = entry.getValue();
-			final List<Instruction> newBranches = new ArrayList<Instruction>(
-					replacements.size());
-			for (final AbstractInsnNode b : replacements) {
-				newBranches.add(instructions.get(b));
+			final Map<AbstractInsnNode, Collection<InstructionBranch>> targets = entry
+					.getValue();
+			int i = 0;
+			for (final Collection<InstructionBranch> list : targets.values()) {
+				i += list.size();
 			}
+			final int[] branches = new int[i];
+			final Instruction[] fromInstructions = new Instruction[i];
+			final int[] fromBranches = new int[i];
+
+			i = 0;
+			int b = 0;
+			for (final Collection<InstructionBranch> list : targets.values()) {
+				for (final InstructionBranch ib : list) {
+					branches[i] = b;
+					fromInstructions[i] = instructions.get(ib.instruction);
+					fromBranches[i] = ib.branch;
+					i++;
+				}
+				b++;
+			}
```

Comment:
I already did not like that the old code was doing allocations merely to pass the data between different interfaces, but the code was small enough, and now it become bigger 😞 

IMO allocations can be avoided and code can be simplified by introduction of "mapper" interface similar to `java.util.function.Function<AbstractInstructionNode, Instruction>`

```
mapper = (AbstractInsnNode node) -> instructions.get(node);
mapper.apply(node).replaceBranches(mapper, targets);
```